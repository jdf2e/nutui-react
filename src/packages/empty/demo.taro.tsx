import React from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'
import Demo1 from './demos/taro/demo1'
import Demo2 from './demos/taro/demo2'
import Demo3 from './demos/taro/demo3'
import Demo4 from './demos/taro/demo4'
import Demo5 from './demos/taro/demo5'
import Demo6 from './demos/taro/demo6'

const EmptyDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      ce5c5446: '基础用法',
      c38a08ee: 'Size 为 small 时，可用于半屏',
      c38a08ef: '自定义内容大小',
      b840c88f: '图片类型，内置3个',
      a74a1fd4: '自定义图片',
      '8dab2f66': '底部内容',
    },
    'zh-TW': {
      ce5c5446: '基礎用法',
      c38a08ee: 'Size 为 small 时，可用于半屏',
      c38a08ef: '自定義內容大小',
      b840c88f: '圖片類型，內置3個',
      a74a1fd4: '自定義圖片',
      '8dab2f66': '底部內容',
    },
    'en-US': {
      ce5c5446: 'Basic usage',
      c38a08ee: 'Size is small',
      c38a08ef: 'Custom content size',
      b840c88f: 'Picture type, built-in 3',
      a74a1fd4: 'Custom image',
      '8dab2f66': 'Bottom content',
    },
  })

  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.ce5c5446}</h2>
        <Demo1 />
        <h2>{translated.c38a08ee}</h2>
        <Demo2 />
        <h2>{translated.c38a08ef}</h2>
        <Demo3 />
        <h2>{translated.b840c88f}</h2>
        <Demo4 />
        <h2>{translated.a74a1fd4}</h2>
        <Demo5 />
        <h2>{translated['8dab2f66']}</h2>
        <Demo6 />
      </div>
    </>
  )
}

export default EmptyDemo
