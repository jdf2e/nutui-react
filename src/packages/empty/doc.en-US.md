# Empty

## Intro

Placeholder prompt when empty

## Install

```tsx
import { Empty } from '@nutui/nutui-react'
```

## Demo

### Basic usage

:::demo

```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty description="No Data" />
  );
};
export default App;
```

:::

### Custom content size

:::demo

```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty description="No Data" imageSize={100} />
  );
};
export default App;
```

:::

### Picture type, 3 built-in

:::demo

```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <div className="show">
      <Empty status="empty" description="No Content" />
      <Empty status="error" description="Load Failed" />
      <Empty status="network" description="No Network" />
    </div>
  );
};
export default App;
```

:::

### Custom image

:::demo

```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty
      description="No Coupon" 
      image={<img src="https://static-ftcms.jd.com/p/files/61a9e3313985005b3958672e.png" alt=""/>}
     />
  );
};
export default App;
```

:::

### Bottom content

:::demo

```tsx
import  React from "react";
import { Empty, Button } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty status="error" description="Failed to load">
      <div style={{marginTop: "10px"}}>
        <Button icon="refresh" type="primary">重试</Button>
      </div>
    </Empty>
  );
};
export default App;
```

:::

## Empty

### Props

| Property | Description | Type | Default |
| --- | --- | --- | --- |
| image | Image type, supports incoming image URLs | `ReactNode` | `-` |
| imageSize | Image size, the unit of number type is px | `number` \| `string` | `-` |
| description | Description text below the image | `ReactNode` | `-` |
| status | The Default error type | `empty` \| `error` \| `network` | `empty` |

## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/component/configprovider).

| Name | Description | Default |
| --- | --- | --- |
| \--nutui-empty-padding | The padding value of the Empty component image | `32px 0` |
| \--nutui-empty-image-size | The size of the Empty component image | `170px` |
| \--nutui-empty-description-margin-top | Empty component image description margin-top value | `4px` |
| \--nutui-empty-description-color | Empty component image description color value | `#666666` |
| \--nutui-empty-description-font-size | Empty component image description font-size value size | `14px` |
| \--nutui-empty-description-line-height | Empty component image description row height | `20px` |
| \--nutui-empty-description-padding | Empty component picture description padding value | `0 40px` |