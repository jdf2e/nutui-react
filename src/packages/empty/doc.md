#  Empty组件

### 介绍

空状态时的占位提示

### 安装

```javascript
// react
import { Empty } from '@nutui/nutui-react'

```


## 代码演示

### 基础用法
:::demo
```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty description="无数据" />
  );
};
export default App;
```
:::

### 自定义内容大小
:::demo
```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty description="无数据" imageSize={100} />
  );
};
export default App;
```
:::

### 图片类型，内置 3 个
:::demo
```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <div className="show">
      <Empty image="empty" description="无内容" />
      <Empty image="error" description="加载失败/错误" />
      <Empty image="network" description="无网络" />
    </div>
  );
};
export default App;
```
:::

### 自定义图片
:::demo
```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty
      description="无优惠券" 
      image={<img src="https://static-ftcms.jd.com/p/files/61a9e3313985005b3958672e.png" alt=""/>}
     />
  );
};
export default App;
```
:::


### 底部内容
:::demo
```tsx
import  React from "react";
import { Empty } from '@nutui/nutui-react';

const App = () => {
  return (
    <Empty image="error" description="加载失败">
        <div style="margin-top: 10px">
            <nut-button icon="refresh" type="primary">重试</nut-button>
        </div>
    </Empty>
  );
};
export default App;
```
:::
## API

### Props

| 参数         | 说明                             | 类型   | 默认值           |
|--------------|----------------------------------|--------|------------------|
| image         | 图片类型，可选值为 error network empty，支持传入图片 URL              | ReactNode       |
| imageSize        | 图片大小，Number 类型单位为 px                         | Number \| String | -       |
| description         | 图片下方的描述文字 | ReactNode | 无数据                |




## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/component/configprovider)。

| 名称 | 默认值 |
| --- | --- |
| --nutui-empty-padding | ` 32px 0` |
| --nutui-empty-image-size | ` 170px` |
| --nutui-empty-description-margin-top | `  4px` |
| --nutui-empty-description-color | `  #666666` |
| --nutui-empty-description-font-size | `  14px` |
| --nutui-empty-description-line-height | `  20px` |
| --nutui-empty-description-padding | `  0 40px` |
