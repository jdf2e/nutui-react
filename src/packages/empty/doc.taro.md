# Empty组件

空状态时的占位提示

## 引入

```tsx
import { Empty } from '@nutui/nutui-react-taro'
```

## 示例代码

### 基础用法

:::demo

<CodeBlock src='taro/demo1.tsx'></CodeBlock>

:::

### Size 为 small 时，可用于半屏

:::demo

<CodeBlock src='taro/demo2.tsx'></CodeBlock>

:::

### 自定义内容大小

:::demo

<CodeBlock src='taro/demo3.tsx'></CodeBlock>

:::

### 图片类型，内置 3 个

:::demo

<CodeBlock src='taro/demo4.tsx'></CodeBlock>

:::

### 自定义图片

> 如果您是京东站内相关项目的开发，我们特意为您提供了一系列的缺省状态的图片链接，您可通过内部群获取。

:::demo

<CodeBlock src='taro/demo5.tsx'></CodeBlock>

:::

### 底部内容

:::demo

<CodeBlock src='taro/demo6.tsx'></CodeBlock>

:::

## Empty

### Props

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| image | 图片,支持传入图片 URL | `ReactNode` | `-` |
| imageSize | 图片大小，number 类型单位为 px | `number` \| `string` | `-` |
| title | 图片下方的标题 | `ReactNode` | `-` |
| description | 图片下方的描述文字 | `ReactNode` | `-` |
| size | 组件整体大小，适配于全屏或半屏 | `small` \| `base` | `base` |
| status | 默认图片错误类型 | `empty` \| `error` \| `network` | `empty` |
| actions | 可用于处理操作的一组数据 | `Array` | `[]` |

## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/component/configprovider)。

| 名称 | 说明 | 默认值 |
| --- | --- | --- |
| \--nutui-empty-padding | Empty组件图片的padding值 | `32px 40px` |
| \--nutui-empty-image-size | Empty组件图片的尺寸大小 | `160px` |
| \--nutui-empty-image-small-size | size 为 small 时，Empty组件图片的尺寸大小 | `120px` |
| \--nutui-empty-title-margin-top | Empty组件图片标题margin-top的值 | `0px` |
| \--nutui-empty-title-margin-top | Empty组件图片标题margin-top的值 | `8px` |
| \--nutui-empty-title-line-height | Empty组件图片标题行高 | `$font-size-base` |
| \--nutui-empty-description-margin-top | Empty组件图片描述margin-top的值 | `4px` |
| \--nutui-empty-description-line-height | Empty组件图片描述行高 | `1.2` |
| \--nutui-empty-background-color | Empty组件背景色 | `#fff` |
