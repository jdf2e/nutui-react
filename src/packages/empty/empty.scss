.nut-empty {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  padding: $empty-padding;
  background-color: $gray6;

  &__image {
    width: $empty-image-size;
    height: $empty-image-size;

    .img {
      width: 100%;
      height: 100%;
    }

    // 兼容小程序标签和img-slot
    img,
    image {
      width: 100%;
      height: 100%;
    }
  }

  &__description {
    margin-top: $empty-description-margin-top;
    padding: $empty-description-padding;
    color: $empty-description-color;
    font-size: $empty-description-font-size;
    line-height: $empty-description-line-height;
  }
}
