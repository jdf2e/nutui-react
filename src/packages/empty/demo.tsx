import React, { useState } from 'react'
import { Refresh } from '@nutui/icons-react'
import { Empty } from './empty'
import { Button } from '@/packages/button/button'
import { Tabs } from '@/packages/tabs/tabs'
import { TabPane } from '@/packages/tabpane/tabpane'
import { useTranslate } from '../../sites/assets/locale'

interface T {
  ce5c5446: string
  c38a08ee: string
  c38a08ef: string
  b840c88f: string
  a74a1fd4: string
  '8dab2f66': string
  cfbdc782: string
  cfbdc784: string
  cfbdc781: string
  c3a3a1d2: string
  e51e4582: string
  adb1a8b2: string
  a52bef0c: string
  d04fcbda: string
  b8a453e3: string
}

const EmptyDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      ce5c5446: '基础用法',
      c38a08ee: 'Size 为 small 时，可用于半屏',
      c38a08ef: '自定义内容大小',
      b840c88f: '图片类型，内置3个',
      a74a1fd4: '自定义图片',
      '8dab2f66': '底部内容',
      cfbdc782: '标题',
      cfbdc784: '操作按钮',
      cfbdc781: '无数据',
      c3a3a1d2: '无内容',
      e51e4582: '加载失败/错误',
      adb1a8b2: '无网络',
      a52bef0c: '无优惠券',
      d04fcbda: '加载失败',
      b8a453e3: '重试',
    },
    'zh-TW': {
      ce5c5446: '基礎用法',
      c38a08ee: 'Size 为 small 时，可用于半屏',
      c38a08ef: '自定義內容大小',
      b840c88f: '圖片類型，內置3個',
      a74a1fd4: '自定義圖片',
      '8dab2f66': '底部內容',
      cfbdc782: '标题',
      cfbdc784: '操作按钮',
      cfbdc781: '無數據',
      c3a3a1d2: '無內容',
      e51e4582: '加載失敗/錯誤',
      adb1a8b2: '無網絡',
      a52bef0c: '無優惠券',
      d04fcbda: '加載失敗',
      b8a453e3: '重試',
    },
    'en-US': {
      ce5c5446: 'Basic usage',
      c38a08ee: 'Size is small',
      c38a08ef: 'Custom content size',
      b840c88f: 'Picture type, built-in 3',
      a74a1fd4: 'Custom image',
      '8dab2f66': 'Bottom content',
      cfbdc782: 'Title',
      cfbdc784: 'Operation Button',
      cfbdc781: 'No Data',
      c3a3a1d2: 'No Content',
      e51e4582: 'Load Failed',
      adb1a8b2: 'No Network',
      a52bef0c: 'No Coupon',
      d04fcbda: 'Failed to load',
      b8a453e3: 'Retry',
    },
  })
  const [tabvalue, setTabvalue] = useState<any>('')

  return (
    <>
      <div className="demo">
        <h2>{translated.ce5c5446}</h2>
        <Empty
          title={translated.cfbdc782}
          description={translated.cfbdc781}
          actions={[
            { text: translated.cfbdc784 },
            { text: translated.cfbdc784 },
          ]}
        />
        <Empty
          description={translated.cfbdc781}
          actions={[{ text: translated.cfbdc784 }]}
          style={{ marginTop: '10px' }}
        />
        <Empty
          description={translated.cfbdc781}
          style={{ marginTop: '10px' }}
        />
        <h2>{translated.c38a08ee}</h2>
        <Empty description={translated.cfbdc781} size="small" />
        <h2>{translated.c38a08ef}</h2>
        <Empty description={translated.cfbdc781} imageSize={80} />
        <h2>{translated.b840c88f}</h2>
        <Tabs
          value={tabvalue}
          onChange={(paneKey) => {
            setTabvalue(paneKey)
          }}
        >
          <TabPane title={translated.c3a3a1d2} value="0">
            <Empty status="empty" title={translated.c3a3a1d2} />
          </TabPane>
          <TabPane title={translated.e51e4582} value="1">
            <Empty status="error" title={translated.e51e4582} />
          </TabPane>
          <TabPane title={translated.adb1a8b2} value="2">
            <Empty status="network" title={translated.adb1a8b2} />
          </TabPane>
        </Tabs>
        <h2>{translated.a74a1fd4}</h2>
        <Empty
          description={translated.a52bef0c}
          image={
            <img
              src="https://static-ftcms.jd.com/p/files/61a9e3313985005b3958672e.png"
              alt=""
            />
          }
        />

        <h2>{translated['8dab2f66']}</h2>
        <Empty
          status="error"
          description={translated.d04fcbda}
          style={{ marginBottom: '20px' }}
        >
          <div style={{ marginTop: '10px' }}>
            <Button icon={<Refresh />} type="primary">
              {translated.b8a453e3}
            </Button>
          </div>
        </Empty>
      </div>
    </>
  )
}

export default EmptyDemo
