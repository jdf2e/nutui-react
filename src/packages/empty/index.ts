import { Empty } from './empty'

export type { EmptyProps } from './empty'

export default Empty
