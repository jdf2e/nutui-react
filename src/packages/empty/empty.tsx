import React, { FunctionComponent, useEffect, useState, ReactNode } from 'react'
import classNames from 'classnames'
import { useConfig } from '@/packages/configprovider'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'
import { EmptyAction } from '@/packages/empty/index'

import Button from '../button'

type statusOptions = {
  [key: string]: string
}
/**
 * 内置图片地址
 */
const defaultStatus: statusOptions = {
  empty:
    'https://storage.360buyimg.com/imgtools/30186cfda0-0d3eee40-c0ac-11ee-9382-9125782aa3b8.png',
  error:
    'https://storage.360buyimg.com/imgtools/f3278d0ebb-0ce360c0-c0ac-11ee-8375-193101bb1a46.png',
  network:
    'https://storage.360buyimg.com/imgtools/43c30f7e29-0d483d10-c0ac-11ee-bec4-eb4d2a09a51d.png',
}
export interface EmptyProps extends BasicComponent {
  image?: ReactNode
  imageSize: number | string
  title: ReactNode
  description: ReactNode
  size: 'small' | 'base'
  status: 'empty' | 'error' | 'network'
  actions: Array<EmptyAction>
}

const defaultProps = {
  ...ComponentDefaults,
  title: '',
  description: '',
  imageSize: '',
  size: 'base',
  status: 'empty',
  actions: [],
} as EmptyProps

const classPrefix = `nut-empty`
export const Empty: FunctionComponent<
  Partial<EmptyProps> & Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
> = (props) => {
  const { locale } = useConfig()
  const {
    image,
    imageSize,
    title,
    description,
    children,
    className,
    size,
    status,
    actions,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }

  const [imgStyle, setImgStyle] = useState<any>({})

  const imageUrl = image || defaultStatus[status]
  const imageNode =
    typeof imageUrl === 'string' ? (
      <img className="img" src={imageUrl} alt="empty" />
    ) : (
      imageUrl
    )

  useEffect(() => {
    setImgStyle(() => {
      if (!imageSize) {
        return {}
      }
      if (typeof imageSize === 'number') {
        return {
          width: `${imageSize}px`,
          height: `${imageSize}px`,
        }
      }
      return {
        width: imageSize,
        height: imageSize,
      }
    })
  }, [imageSize])
  const classes = classNames({
    [`${classPrefix}-${size}`]: size !== 'base',
  })
  const cls = classNames(classPrefix, classes, className)

  return (
    <div className={cls} {...rest}>
      <div className={`${classPrefix}-image`} style={imgStyle}>
        {imageNode}
      </div>
      {typeof title === 'string' && title ? (
        <div className={`${classPrefix}-title`}>{title}</div>
      ) : (
        title
      )}
      {typeof description === 'string' ? (
        <div className={`${classPrefix}-description`}>{description}</div>
      ) : (
        description
      )}
      {actions.length > 0 && (
        <div className={`${classPrefix}-actions`}>
          {actions.map((item, index) => {
            return (
              <Button
                className={classNames({
                  [`${classPrefix}-actions-right`]: actions.length === 1,
                  [`${classPrefix}-actions-left`]:
                    actions.length > 1 && index === 0,
                })}
                type={`${
                  actions.length > 1 && index === 0 ? 'default' : 'primary'
                }`}
                size="small"
                fill="outline"
                key={`action-${index}`}
              >
                {item?.text}
              </Button>
            )
          })}
        </div>
      )}
      {children}
    </div>
  )
}

Empty.defaultProps = defaultProps
Empty.displayName = 'NutEmpty'
