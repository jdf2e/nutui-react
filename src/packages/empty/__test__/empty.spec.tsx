import * as React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import { Empty } from '../empty'

test('should render description correctly', () => {
  const { container, getByText } = render(<Empty description="暂无数据" />)
  expect(getByText('暂无数据')).toBeTruthy()
  expect(container).toMatchSnapshot()
})
test('should render imageSize correctly', () => {
  const { container } = render(<Empty description="暂无数据" imageSize={100} />)
  expect(container.querySelector('.nut-empty-image')).toHaveStyle({
    width: '100px',
  })
})
test('should render image props correctly', () => {
  const { container } = render(<Empty status="empty" />)
  expect(container.querySelector('.img')).toHaveAttribute(
    'src',
    'https://storage.360buyimg.com/imgtools/30186cfda0-0d3eee40-c0ac-11ee-9382-9125782aa3b8.png'
  )
})
test('should render actions correctly', () => {
  const { container, getByTestId } = render(
    <Empty
      data-testid="testEmpty"
      title="标题"
      description="暂无数据"
      actions={[{ text: '操作1' }, { text: '操作2' }]}
    />
  )
  expect(container.querySelector('.nut-empty-actions')).toBeTruthy()
  expect(
    container.querySelectorAll('.nut-empty-actions .nut-button').length
  ).toEqual(2)
})
test('should render custom image correctly', () => {
  const { getByTestId } = render(
    <Empty
      description="店铺为空"
      image={
        <img
          src="https://static-ftcms.jd.com/p/files/61a9e3313985005b3958672e.png"
          alt=""
          data-testid="testImg"
        />
      }
    />
  )
  expect(getByTestId('testImg')).toHaveAttribute(
    'src',
    'https://static-ftcms.jd.com/p/files/61a9e3313985005b3958672e.png'
  )
})
