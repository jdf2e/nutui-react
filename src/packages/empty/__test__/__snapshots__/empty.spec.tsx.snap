// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render description correctly 1`] = `
<div>
  <div
    class="nut-empty"
  >
    <div
      class="nut-empty-image"
    >
      <img
        alt="empty"
        class="img"
        src="https://storage.360buyimg.com/imgtools/30186cfda0-0d3eee40-c0ac-11ee-9382-9125782aa3b8.png"
      />
    </div>
    <div
      class="nut-empty-description"
    >
      暂无数据
    </div>
  </div>
</div>
`;
